Mathlib/Tactic/Tauto.lean

```diff
@@ -26,7 +26,11 @@ open Qq
 initialize registerTraceClass `tauto
 
 /-- Tries to apply de-Morgan-like rules on a hypothesis. -/
-def distribNotAt (h : LocalDecl) (g : MVarId) : MetaM MVarId := do
+def distribNotAt (fvarUserName : Name) (g : MVarId) : MetaM MVarId := g.withContext do
+  let ctx ← getLCtx
+  let h ← match LocalContext.findFromUserName? ctx fvarUserName with
```

Comment:
Using user names is flaky (we might have two hypotheses with the same user name, after all).  This function should probably return the new fvar id (which you can get from the output of `replace`).

Reply:
If I use `FVarId`s, then I don't see an easy way to keep track of the remaining hypotheses to work on (the `for h in ← getLCtx` in `distribNot`). We'll run `distribNotAt` on the first `FVarId`, and then by the time we get around to the second `FVarId`, it might no longer be present in the local context! I suppose that I can the information in `AssertAfterResult.subst` to thread that information through?

Reply:
Ok, done. Now we key hypothesis by their fvar ID, and we keep track of renamings by passing around the list of what remains to do. This makes things a bit more complex, but is probably worth it for the increased robustness.